### PR TITLE
Emit ansi color codes in the `rendered` field of json diagnostics

### PR DESCRIPTION
**Pull request by oli-obk**
cc @ljedrz 

Implemented for https://github.com/rust-lang/rust/pull/56595#issuecomment-447645115 (x.py clippy)

### CODE REVIEW

**Comment by ehuss** [comment]
I've been looking at doing something similar to this.  I'm considering using JSON output only for Cargo for a few reasons.  Do you happen to know if the "rendered" field is fully consistent with the normal output?

**Comment by oli-obk** [comment]
> Do you happen to know if the "rendered" field is fully consistent with the normal output?

Yes it's identical. It's actually generated by invoking the regular diagnostic renderer and just recording the output.

**Comment by eddyb** [comment]
@oli-obk Can we get `compiletest` to use rendered stuff, too? Right now it's dumping JSON out, which is pretty bad, but it would be even worse after this.

**Comment by oli-obk** [comment]
Oh yea, totally forgot about this. I've been meaning to fix that in forever.

I'll do that in this PR as to not regress the status quo even further.

**Comment by mark-i-m** [comment]
My email is telling me I was r?-ed, but I can't see myself in this thread... am I missing something?

**Comment by oli-obk** [comment]
the `r?` didn't work, so I deleted the comment.

**Review by mark-i-m — APPROVED** [review]
LGTM, even though I don't have r? permissions :)

**Comment by oli-obk** [comment]
> Right now it's dumping JSON out, which is pretty bad, but it would be even worse after this.

Note that this would not have changed the rendered output (because that would severely screw with `.stderr` and `.stdout` files). But I fixed the problem anyway.

r? @eddyb

File: src/tools/compiletest/src/json.rs [hunk 1]
```diff
@@ -62,19 +62,20 @@ struct DiagnosticCode {
...
                     Ok(diagnostic) => diagnostic.rendered,
                     Err(error) => {
-                        proc_res.fatal(Some(&format!(
+                        print!(
                             "failed to decode compiler output as json: \
                              `{}`\nline: {}\noutput: {}",
                             error, line, output
-                        )));
+                        );
+                        panic!()
```
Comment (eddyb) [inline src/tools/compiletest/src/json.rs]: Does this `panic!` match what `proc_res.fatal` would do?

File: src/tools/compiletest/src/runtest.rs [hunk 2]
```diff
@@ -3290,7 +3250,9 @@ impl ProcRes {
              {}\n\
              ------------------------------------------\n\
              \n",
-            self.status, self.cmdline, self.stdout, self.stderr
+            self.status, self.cmdline,
+            json::extract_rendered(&self.stdout),
+            json::extract_rendered(&self.stderr),
```
Comment (eddyb) [inline src/tools/compiletest/src/runtest.rs]: I assume this is the bulk of the changes? :tada:

Comment (oli-obk) [reply src/tools/compiletest/src/runtest.rs]: jop

File: src/test/ui/lint/use_suggestion_json.rs [hunk 3]
```diff
@@ -1,5 +1,5 @@
 // ignore-cloudabi
-// compile-flags: --error-format pretty-json -Zunstable-options
+// compile-flags: --error-format pretty-json -Zunstable-options --colorful-json=true
```
Comment (estebank) [inline src/test/ui/lint/use_suggestion_json.rs]: I don't see a test for the _current_ output here :-/

Also, although I see the immediate benefit of this, I would like for us to tackle a more "general" solution at some point where we can also produce HTML, for example.

Comment (oli-obk) [reply src/test/ui/lint/use_suggestion_json.rs]: We do have other `--error-format pretty-json` ui tests.

So you're imagining something like `--json-rendered=(plain|termcolor|html)`?

Comment (oli-obk) [reply src/test/ui/lint/use_suggestion_json.rs]: ping @estebank

Comment (estebank) [reply src/test/ui/lint/use_suggestion_json.rs]: In my mind something like `--json-rendered=extended` where the textual output would be the same and we'd have extra json fields annotating those with extra information:

```
{"text": "path `std::string::String`", "annotations": [{"start": 6, "end": 25, "highlight_type": "highlighted"}]}
```

That way any tool can consume these and turn them into whatever their needs may be.

That being said, having specific support for both terminal and html output make sense to me, given they probably cover 80% of all likely use cases.

Comment (oli-obk) [reply src/test/ui/lint/use_suggestion_json.rs]: The `--json-rendered` list can be extended in the future to support schemes like the one you proposed. I will change the PR to only support `plain` and `termcolor` for now.

**Comment by oli-obk** [comment]
@estebank I implemented such a command line flag in 7a24be5

We can extend it in the future with other rendering schemes. It's still unstable so not a problem to change significantly anyway

**Comment by Centril** [comment]
Ping from triage, @oli-obk -- you have some test failures.

**Comment by estebank** [comment]
It seems to be a difference in the color of the cli output.

**Comment by oli-obk** [comment]
I found the problem. I made a wrong assumption about how rendering to buffers should handle colors.

@bors r=mark-i-m,eddyb

**Comment by oli-obk** [comment]
I guess color codes differ between windows and linux?

**Pull request merged**